core-bundle/src/Security/Authentication/Token/TokenChecker.php

```diff
@@ -172,4 +198,39 @@ private function getTokenFromSession(string $sessionKey): ?TokenInterface
 
         return $token;
     }
+
+    private function isValidPreviewLink(FrontendPreviewToken $token): bool
+    {
+        $id = $token->getPreviewLinkId();
+
+        if (!isset($this->previewLinks[$id])) {
+            $this->previewLinks[$id] = $this->connection->executeQuery(
+                "SELECT url, showUnpublished, restrictToUrl FROM tl_preview_link
+                WHERE id = :id AND published = '1' AND expiresAt > UNIX_TIMESTAMP()",
+                ['id' => $id],
+            )->fetchAssociative();
+        }
+
+        $previewLink = $this->previewLinks[$id];
+
+        if (!$previewLink) {
+            return false;
+        }
+
+        if ((bool) $previewLink['showUnpublished'] !== $token->showUnpublished()) {
+            return false;
+        }
+
+        if (!$previewLink['restrictToUrl']) {
+            return true;
+        }
+
+        $request = $this->requestStack->getMainRequest();
+
+        if ($request && strtok($request->getUri(), '?') === strtok($previewLink['url'], '?')) {
```

Comment:
What is the correct way to compare two URLs, I should probably use something like `Request::create($previewLink['url'])` here and compare the two objects?

Reply:
I think your solution is fine.

Reply:
Improved in 1d57cdcac1f0e31980100a235af96610b00a3c3e